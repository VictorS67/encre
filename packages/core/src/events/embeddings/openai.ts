import {
  OpenAI as OpenAIClient,
  ClientOptions as OpenAIClientOptions,
} from 'openai';
import type { RequestOptions as OpenAIClientRequestOptions } from 'openai/core';
import { type SecretFields, type SerializedFields } from '../../load/keymap.js';
import { getEnvironmentVariables } from '../../utils/environment.js';
import { type EmbedResult } from '../output/provide/index.js';
import {
  BaseEmbeddings,
  type BaseEmbeddingsCallOptions,
  type BaseEmbeddingsParams,
} from './base.js';

/**
 * Parameters for the OpenAI embeddings, including model and API key specifics.
 */
export interface OpenAIEmbeddingsParams extends BaseEmbeddingsParams {
  /**
   * ID of the model to use. You can use the [List models]({@link https://platform.openai.com/docs/api-reference/models/list}}
   * API to see all of your avaliable models, or see our
   * [Model overview]({@link https://platform.openai.com/docs/models/overview}) for description of them.
   */
  modelName: string;

  /**
   * The number of dimensions the resulting output embeddings should have.
   * Only supported in `text-embedding-3` and later models.
   */
  dimensions?: number;

  /**
   * A unique identifier representing your end-user, which can help OpenAI to monitor
   * and detect abuse.
   *
   * [Learn more]({@link https://platform.openai.com/docs/guides/safety-best-practices})
   */
  user?: string;

  /**
   * API key to use when making requests to OpenAI. Defaults to the value of
   * `OPENAI_API_KEY` environment variable.
   */
  openAIApiKey?: string;

  /**
   * Timeout for the API requests.
   */
  timeout?: number;
}

/**
 * Call options for OpenAI embeddings, including format and additional request options.
 */
export interface OpenAIEmbeddingsCallOptions extends BaseEmbeddingsCallOptions {
  /**
   * Format of the encoding for the embeddings, either as 'float' or 'base64'.
   */
  encondingFormat?: 'float' | 'base64';

  /**
   * Additional options to pass to the underlying axios request.
   */
  options?: OpenAIClientRequestOptions;
}

/**
 * Class to handle embeddings via the OpenAI API, extending the base embeddings functionality.
 *
 * @example
 * ```typescript
 * const embeddings = new OpenAIEmbeddings({ modelName: 'text-embedding-ada-002' });
 * const options = { encondingFormat: 'float' };
 * const response = await embeddings.invoke('Hello!', options);
 * console.log(response);
 * ```
 */
export class OpenAIEmbeddings<
    CallOptions extends
      OpenAIEmbeddingsCallOptions = OpenAIEmbeddingsCallOptions,
  >
  extends BaseEmbeddings<CallOptions>
  implements OpenAIEmbeddingsParams
{
  _isSerializable = true;

  get _secrets(): SecretFields | undefined {
    return {
      openAIApiKey: 'OPENAI_API_KEY',
      organization: 'OPENAI_ORGANIZATION',
    };
  }

  get _aliases(): SerializedFields | undefined {
    return {
      modelName: 'model',
      openAIApiKey: 'openai_api_key',
    };
  }

  static _name(): string {
    return 'OpenAI';
  }

  _embeddingsType(): string {
    return 'openai';
  }

  /**
   * ID of the model to use. You can use the [List models]({@link https://platform.openai.com/docs/api-reference/models/list}}
   * API to see all of your avaliable models, or see our
   * [Model overview]({@link https://platform.openai.com/docs/models/overview}) for description of them.
   */
  modelName = 'text-embedding-ada-002';

  /**
   * The number of dimensions the resulting output embeddings should have.
   * Only supported in `text-embedding-3` and later models.
   */
  dimensions?: number | undefined;

  /**
   * A unique identifier representing your end-user, which can help OpenAI to monitor
   * and detect abuse.
   *
   * [Learn more]({@link https://platform.openai.com/docs/guides/safety-best-practices})
   */
  user?: string | undefined;

  /**
   * The max number of concurrent calls that can be made.
   * Defaults to 2.
   */
  maxConcurrency?: number | undefined;

  /**
   * The max number of retries that can be made for a single call.
   */
  maxRetries?: number | undefined;

  /**
   * Timeout for the API requests.
   */
  timeout?: number;

  /**
   * API key to use when making requests to OpenAI. Defaults to the value of
   * `OPENAI_API_KEY` environment variable.
   */
  openAIApiKey?: string;

  /**
   * Identifier for organization sometimes used in API requestsã€‚
   */
  organization?: string;

  /**
   * OpenAI API Client.
   * @internal
   */
  private _client: OpenAIClient;

  /**
   * OpenAI API request options.
   * @internal
   */
  private _clientOptions: OpenAIClientOptions;

  constructor(
    fields?: Partial<OpenAIEmbeddingsParams> & {
      configuration?: OpenAIClientOptions;
    }
  ) {
    fields = {
      modelName: fields?.modelName ?? 'text-embedding-ada-002',
      maxConcurrency: fields?.maxConcurrency ?? 2,
      maxRetries: fields?.maxRetries,
      dimensions: fields?.dimensions,
      user: fields?.user,
      timeout: fields?.timeout,
      ...fields,
    };

    super(fields);

    this.openAIApiKey =
      fields?.openAIApiKey ?? getEnvironmentVariables('OPENAI_API_KEY');

    if (!this.openAIApiKey) {
      throw new Error('OpenAI API Key not found');
    }

    this.modelName = fields?.modelName ?? this.modelName;

    this.maxConcurrency = fields?.maxConcurrency ?? 2;
    this.maxRetries = fields?.maxRetries;
    this.dimensions = fields?.dimensions;
    this.user = fields?.user;
    this.timeout = fields?.timeout;

    this.organization =
      fields?.configuration?.organization ??
      getEnvironmentVariables('OPENAI_ORGANIZATION');

    this._clientOptions = {
      apiKey: this.openAIApiKey,
      organization: this.organization,
      baseURL: fields?.configuration?.baseURL,
      dangerouslyAllowBrowser: true,
      defaultHeaders: fields?.configuration?.defaultHeaders,
      defaultQuery: fields?.configuration?.defaultQuery,
      ...fields?.configuration,
    };
  }

  /**
   * Retrieves the parameters required for creating an embedding with the OpenAI API.
   * This method extracts and prepares the parameters based on the model configurations
   * and any additional options specified in the call.
   *
   * @param options Optional serialized call options which may override default model settings.
   * @returns An object containing the necessary parameters for the OpenAI API request,
   *          excluding the 'input' parameter which is handled separately.
   */
  getParams(
    options?: this['SerializedCallOptions']
  ): Omit<OpenAIClient.EmbeddingCreateParams, 'input'> {
    return {
      model: this.modelName,
      dimensions: this.dimensions,
      encoding_format: options?.encondingFormat,
      user: this.user,
    };
  }

  _identifyParams(): Omit<OpenAIClient.EmbeddingCreateParams, 'input'> & {
    model_name: string;
  } & OpenAIClientOptions {
    return {
      model_name: this.modelName,
      ...this.getParams(),
      ...this._clientOptions,
    };
  }

  /**
   * Embeds a document using the OpenAI API.
   * @param document The document to embed.
   * @param options Serialized call options.
   * @returns A promise resolving to the embedding result.
   * @internal
   * @example
   * ```typescript
   * const embeddings = new OpenAIEmbeddings({ modelName: 'text-embedding-ada-002' });
   * const result = await embeddings._embed("Hello, world!", {});
   * console.log(result.embedding);
   * ```
   */
  async _embed(
    document: string,
    options: this['SerializedCallOptions']
  ): Promise<EmbedResult> {
    const params = this.getParams(options);

    const response = await this.completionWithRetry(
      {
        ...params,
        input: document,
      },
      {
        signal: options?.signal,
        ...options?.options,
      }
    );

    return {
      embedding: response.data[0].embedding,
      embedOutput: {
        tokenUsage: {
          promptTokens: response.usage.prompt_tokens,
          totalTokens: response.usage.total_tokens,
        },
      },
    };
  }

  /**
   * Handles retries for the embedding creation in case of failures.
   * @param request The embedding creation parameters.
   * @param options Optional client request configurations.
   * @returns A promise resolving to the embedding creation response.
   * @example
   * ```typescript
   * const embeddings = new OpenAIEmbeddings({ modelName: 'text-embedding-ada-002' });
   * const requestParams = { input: "Example document", model: "text-embedding-ada-002" };
   * const response = await embeddings.completionWithRetry(requestParams);
   * console.log(response.data[0].embedding);
   * ```
   */
  async completionWithRetry(
    request: OpenAIClient.EmbeddingCreateParams,
    options?: OpenAIClientRequestOptions
  ): Promise<OpenAIClient.Embeddings.CreateEmbeddingResponse> {
    const requestOptions: OpenAIClientRequestOptions =
      this._getRequestOptions(options);
    return this.caller.call(async () => {
      const res = await this._client.embeddings.create(request, requestOptions);
      return res;
    });
  }

  /**
   * Builds request options for the API call, merging default client options with
   * method-specific options.
   * @param options Optional additional request configurations.
   * @returns The combined request options.
   * @internal
   */
  private _getRequestOptions(
    options?: OpenAIClientRequestOptions
  ): OpenAIClientRequestOptions {
    if (!this._client) {
      const params = {
        ...this._clientOptions,
        timeout: this.timeout,
        maxRetries: 0,
      };

      if (!params.baseURL) {
        delete params.baseURL;
      }

      this._client = new OpenAIClient(params);
    }

    const requestOptions: OpenAIClientRequestOptions = {
      ...this._clientOptions,
      ...options,
    };

    return requestOptions;
  }
}
