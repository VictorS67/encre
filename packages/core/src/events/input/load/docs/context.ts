/**
 * Defines the structure for inputs used to create a readable context. This interface is for feeding page
 * content and associated metadata into processing workflows.
 *
 * @template Metadata The type for metadata associated with the context, allowing for custom data structures.
 */
export interface ContextInput<Metadata = Record<string, unknown>> {
  /**
   * The main content of the page, typically as a string.
   */
  pageContent: string;

  /**
   * Optional metadata associated with the page content, providing additional information or annotations.
   */
  metadata?: Metadata;
}

/**
 * Represents a single readable context derived from a page of content. This class encapsulates the content
 * and metadata, providing a structured way to handle such information within processing or analytical operations.
 *
 * @template Metadata The type for metadata that augments the page content with additional data or annotations.
 */
export class Context<Metadata = Record<string, unknown>>
  implements ContextInput<Metadata>
{
  /**
   * The content of the page stored as a string.
   */
  pageContent = '';

  /**
   * Metadata associated with the page content, of type Metadata.
   */
  metadata: Metadata;

  /**
   * Constructs a new instance of Context, initializing it with specified page content and optional metadata.
   * @param fields An object containing pageContent and optionally metadata, conforming to ContextInput.
   */
  constructor(fields: ContextInput<Metadata>) {
    this.pageContent = fields.pageContent
      ? fields.pageContent.toString()
      : this.pageContent;
    this.metadata = fields.metadata ?? ({} as Metadata);
  }

  /**
   * Static method to check if a given value is an instance of Context.
   * @param value The value to check.
   * @returns True if the value is an instance of Context, false otherwise.
   */
  static isContext(value: unknown): value is Context {
    return value instanceof Context;
  }
}

/**
 * Represents types that can be treated as a Context. This includes single Context instances,
 * arrays of Context, or strings that could potentially be converted into a Context.
 */
export type ContextLike = Context | Context[] | string;

/**
 * Determines whether a given value qualifies as a ContextLike. This function is used to validate
 * if a value can either be directly used as a Context, converted to one, or is an array of Contexts.
 *
 * @param value The value to evaluate.
 * @returns True if the value is a string, a Context instance, or an array of Context instances; false otherwise.
 */
export function isContextLike(value: unknown): value is ContextLike {
  if (Array.isArray(value)) {
    return value.every((v) => Context.isContext(v));
  }

  return typeof value === 'string' || Context.isContext(value);
}
