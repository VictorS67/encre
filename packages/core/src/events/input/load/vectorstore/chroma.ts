import {
  ChromaClient,
  Collection as ChromaCollection,
  CollectionMetadata as ChromaCollectionMetadata,
  Metadata as ChromaMetadata,
  IncludeEnum,
  QueryResponse,
  Where,
} from 'chromadb';
import { v4 as uuid } from 'uuid';

import { maximalMarginalRelevance } from '../../../../utils/math.js';
import { Context } from '../docs/index.js';
import { BaseVectorStore, type BaseVectorStoreField } from './base.js';

/**
 * Defines the fields for the ChromaVectorStore class.
 *
 * @example
 * ```typescript
 * const vectorStoreConfig: ChromaVectorStoreField = {
 *   url: 'my-chroma-instance-url',
 *   numDimensions: 128,
 *   collectionName: 'myCollection',
 *   collectionMetadata: { createdBy: 'Data Science Team' },
 *   filter: {
 *     where: { field: 'category', value: 'technology', operation: 'EQUAL' },
 *     includeEmbeddings: true
 *   }
 * };
 * ```
 */
export interface ChromaVectorStoreField extends BaseVectorStoreField {
  /**
   * Base URL of the Chroma backend service.
   */
  url: string;

  /**
   * Number of dimensions for each embedding vector.
   */
  numDimensions?: number;

  /**
   * Name of the collection within Chroma where vectors are stored.
   */
  collectionName?: string;

  /**
   * Additional metadata related to the collection.
   */
  collectionMetadata?: ChromaCollectionMetadata;

  /**
   * Optional filter settings that can be applied during search operations.
   */
  filter?: ChromaVectorStoreFilterType;
}

/**
 * Specifies the parameters for deleting vectors from a Chroma vector store.
 *
 * @example
 * ```typescript
 * // Example of deleting specific vectors by IDs
 * const deleteParams: ChromaDeleteParams<ChromaVectorStoreFilterType> = {
 *   ids: ['id1', 'id2']
 * };
 *
 * // Example of deleting vectors by a custom filter
 * const deleteParamsWithFilter: ChromaDeleteParams<ChromaVectorStoreFilterType> = {
 *   filter: {
 *     where: { field: 'age', value: 25, operation: 'LESS_THAN' }
 *   }
 * };
 * ```
 */
export interface ChromaDeleteParams<T> {
  /**
   * Optional. Array of vector IDs to delete.
   */
  ids?: string[];

  /**
   * Optional. A filter defining which vectors to delete based on their metadata and content.
   */
  filter?: T;
}

/**
 * Defines the filter type used in ChromaVectorStore operations to refine search and deletion operations.
 *
 * @example
 * ```typescript
 * // Example filter to include embeddings and filter based on metadata
 * const filter: ChromaVectorStoreFilterType = {
 *   where: { field: 'status', value: 'active', operation: 'EQUAL' },
 *   includeEmbeddings: true
 * };
 * ```
 */
export type ChromaVectorStoreFilterType = {
  /**
   * Optional. Specifies conditions used to filter vectors based on their metadata.
   */
  where?: Where;

  /**
   * Optional. Indicates whether to include vector embeddings in the results.
   */
  includeEmbeddings?: boolean;
};

/**
 * The ChromaVectorStore class provides an in-memory vector store that integrates
 * with a Chroma backend service for storing and retrieving vector embeddings associated with content.
 * It supports operations like adding vectors, deleting vectors based on IDs or filters, and conducting
 * similarity and maximal marginal relevance searches.
 *
 * @example
 * ```typescript
 * // Instantiate the ChromaVectorStore
 * const vectorStore = new ChromaVectorStore({
 *   url: 'my-chroma-instance-url',
 *   collectionName: 'myVectorCollection'
 * });
 *
 * // Example embeddings and associated contexts
 * const embeddings = [
 *   [0.1, 0.2, 0.3],
 *   [0.4, 0.5, 0.6]
 * ];
 * const contexts = [
 *   new Context({ pageContent: "Content about AI", metadata: { topic: "AI" }}),
 *   new Context({ pageContent: "Content about ML", metadata: { topic: "ML" }})
 * ];
 *
 * // Add vectors to the store
 * await vectorStore.addVectors(embeddings, contexts);
 *
 * // Perform a similarity search
 * const queryEmbedding = [0.15, 0.25, 0.35];
 * const topK = 1;
 * const similarityResults = await vectorStore.similaritySearch(queryEmbedding, topK);
 *
 * // Display the results
 * console.log("Similarity Search Results:", similarityResults.map(r => ({
 *   content: r[0].pageContent,
 *   similarityScore: r[1]
 * })));
 *
 * // Optionally, delete vectors
 * await vectorStore.deleteVectors({ ids: [similarityResults[0][0].id] });
 * ```
 */
export class ChromaVectorStore
  extends BaseVectorStore
  implements ChromaVectorStoreField
{
  declare FilterType: ChromaVectorStoreFilterType;

  _isSerializable = true;

  /**
   * Base URL of the Chroma backend service.
   */
  url: string;

  /**
   * Number of dimensions for each embedding vector.
   */
  numDimensions?: number | undefined;

  /**
   * Name of the collection within Chroma where vectors are stored.
   */
  collectionName?: string | undefined;

  /**
   * Additional metadata related to the collection.
   */
  collectionMetadata?: ChromaCollectionMetadata | undefined;

  /**
   * Optional filter settings that can be applied during search operations.
   */
  filter?: ChromaVectorStoreFilterType | undefined;

  /**
   * Chroma API Client.
   * @internal
   */
  private _client: ChromaClient;

  _vectorstoreType(): string {
    return 'chroma';
  }

  constructor(fields?: Partial<ChromaVectorStoreField>) {
    super(fields);

    this.url = fields?.url ?? 'http://localhost:8000';
    this.numDimensions = fields?.numDimensions;
    this.collectionName = fields?.collectionName;
    this.collectionMetadata = fields?.collectionMetadata;
    this.filter = fields?.filter;
  }

  async addVectors(
    embeddings: number[][],
    context: Context[],
    options?: { ids?: string[] }
  ): Promise<void> {
    if (embeddings.length === 0) {
      return;
    }

    if (embeddings.length !== context.length) {
      throw new Error(
        'Chroma addVectors: unmatched # of embeddings and # of contexts'
      );
    }

    if (this.numDimensions !== embeddings[0].length) {
      console.warn(
        `Chroma addVectors: unmatched # of dimensions and # of embeddings. Now change dimensions to ${embeddings[0].length}`
      );

      this.numDimensions = embeddings[0].length;
    }

    let ids: string[] = options?.ids ?? [];
    if (ids.length > 0 && ids.length > embeddings.length) {
      console.warn(
        'Chroma addVectors: # of ids is larger than # of embeddings. Now truncate last few ids'
      );

      ids = ids.slice(0, embeddings.length);
    } else if (ids.length > 0 && ids.length < embeddings.length) {
      console.warn(
        'Chroma addVectors: # of ids is less than # of embeddings. Now pad last few ids with some random ids'
      );

      const paddingArray = Array.from(
        { length: embeddings.length - ids.length },
        () => uuid()
      );
      ids = ids.concat(paddingArray);
    } else if (ids.length === 0) {
      ids = Array.from({ length: embeddings.length }, () => uuid());
    }

    const collection: ChromaCollection = await this._getCollection();

    const metadatas = context.map(({ metadata }) => {
      let locFrom: number | undefined;
      let locTo: number | undefined;

      if (
        metadata?.loc &&
        typeof metadata.loc === 'object' &&
        metadata.loc !== null &&
        !Array.isArray(metadata.loc) &&
        'lines' in metadata.loc
      ) {
        if (
          typeof metadata.loc.lines === 'object' &&
          metadata.loc.lines !== null &&
          !Array.isArray(metadata.loc.lines) &&
          'from' in metadata.loc.lines &&
          typeof metadata.loc.lines.from === 'number'
        ) {
          locFrom = metadata.loc.lines.from;
        }

        if (
          typeof metadata.loc.lines === 'object' &&
          metadata.loc.lines !== null &&
          !Array.isArray(metadata.loc.lines) &&
          'to' in metadata.loc.lines &&
          typeof metadata.loc.lines.to === 'number'
        ) {
          locTo = metadata.loc.lines.to;
        }
      }

      const newMetadata: Context['metadata'] = {
        ...metadata,
        ...(locFrom !== undefined && { locFrom }),
        ...(locTo !== undefined && { locTo }),
      };

      if (newMetadata.loc) {
        delete newMetadata.loc;
      }

      return newMetadata;
    }) as ChromaMetadata[];

    await collection.upsert({
      ids,
      embeddings,
      metadatas,
      documents: context.map(({ pageContent }) => pageContent),
    });
  }

  async deleteVectors(params?: ChromaDeleteParams<this['FilterType']>) {
    const collection = await this._getCollection();

    if (params?.ids && params?.filter?.where) {
      await collection.delete({
        ids: params.ids,
        where: { ...params.filter.where },
      });
    } else if (params?.ids) {
      await collection.delete({ ids: params.ids });
    } else if (params?.filter?.where) {
      await collection.delete({
        where: { ...params.filter.where },
      });
    } else {
      throw new Error(
        "Chroma deleteVectors: must provide either 'ids' or 'filter'"
      );
    }
  }

  async similaritySearch(
    query: number[],
    topK: number,
    filter?: this['FilterType']
  ): Promise<[Context, number][]> {
    const _filter: this['FilterType'] | undefined = filter ?? this.filter;

    const collection: ChromaCollection = await this._getCollection();
    const include: IncludeEnum[] = [
      IncludeEnum.Documents,
      IncludeEnum.Embeddings,
      IncludeEnum.Distances,
      IncludeEnum.Metadatas,
    ];

    let result: QueryResponse;
    if (_filter?.where) {
      result = await collection.query({
        queryEmbeddings: query,
        nResults: topK,
        where: { ..._filter.where },
        include,
      });
    } else {
      result = await collection.query({
        queryEmbeddings: query,
        nResults: topK,
        include,
      });
    }

    const { ids, distances, embeddings, documents, metadatas } = result;
    if (!ids || !distances || !embeddings || !documents || !metadatas) {
      return [];
    }

    const [firstIds] = ids;
    const [firstDistances] = distances;
    const [firstEmbeddings] = embeddings;
    const [firstDocuments] = documents;
    const [firstMetadatas] = metadatas;

    const results: [Context, number][] = [];
    for (let i = 0; i < firstIds.length; i++) {
      let metadata: Context['metadata'] = firstMetadatas?.[i] ?? {};

      if (metadata.locFrom && metadata.locTo) {
        metadata = {
          ...metadata,
          loc: {
            lines: {
              from: metadata.locFrom,
              to: metadata.locTo,
            },
          },
        };

        delete metadata.locFrom;
        delete metadata.locTo;
      }

      if (_filter?.includeEmbeddings) {
        metadata.embedding = firstEmbeddings?.[i];
      }

      results.push([
        new Context({
          pageContent: firstDocuments?.[i] ?? '',
          metadata,
        }),
        firstDistances[i],
      ]);
    }

    return results;
  }

  async maxMarginalRelevanceSearch(
    query: number[],
    topK: number,
    lambda: number,
    filter?: this['FilterType'] | undefined
  ): Promise<[Context, number][]> {
    const _filter = filter ?? this.filter;
    const includeEmbeddingsFlag = _filter?.includeEmbeddings || false;

    const searches = await this.similaritySearch(query, topK, {
      ..._filter,
      includeEmbeddings: true,
    });

    const embeddings = searches.map(
      (s) => s[0].metadata.embedding
    ) as number[][];

    const [mmrIndexes, mmrSimilarities] = maximalMarginalRelevance(
      query,
      embeddings,
      lambda,
      topK
    );

    return mmrIndexes.map((index) => {
      const context = searches[index][0];
      const similarity = mmrSimilarities[index];

      if (!includeEmbeddingsFlag) {
        delete context.metadata.embedding;
      }

      return [context, similarity];
    });
  }

  /**
   * Retrieves a collection from the Chroma backend. If the collection does not exist, it creates a new one using
   * the specified collection name and optional metadata. This method ensures that the required collection is available
   * for storing and retrieving vector embeddings.
   *
   * @returns A promise that resolves to the ChromaCollection object, allowing for further interaction with the collection.
   * @throws An error if the Chroma service fails to create or retrieve the collection.
   * @internal
   * @example
   * ```typescript
   * const vectorStore = new ChromaVectorStore({
   *   url: 'my-chroma-instance-url',
   *   collectionName: 'myVectorCollection'
   * });
   * try {
   *   const collection = await vectorStore._getCollection();
   *   console.log('Collection ready for use:', collection);
   * } catch (error) {
   *   console.error('Failed to access or create the collection:', error);
   * }
   * ```
   */
  async _getCollection(): Promise<ChromaCollection> {
    if (!this._client) {
      this._client = new ChromaClient({
        path: this.url,
      });
    }

    try {
      const collection: ChromaCollection =
        await this._client.getOrCreateCollection({
          name: this._getCollectionName(),
          ...(this.collectionMetadata && { metadata: this.collectionMetadata }),
        });

      return collection;
    } catch (err) {
      throw new Error(`Chroma failed to getOrCreateCollection: ${err}`);
    }
  }

  /**
   * Generates a unique collection name based on the `collectionName` field. If `collectionName` is not set,
   * it generates a random UUID as the collection name. This method ensures a consistent naming strategy for
   * collections used within the Chroma service.
   *
   * @returns The name of the collection as a string.
   * @internal
   */
  private _getCollectionName(): string {
    if (!this.collectionName) {
      return `encre-chroma-${uuid()}`;
    }

    return this.collectionName;
  }
}
