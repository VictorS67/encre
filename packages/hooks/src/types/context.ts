export interface Context {
  pageContent: string;
  metadata?: Record<string, unknown>;
}
