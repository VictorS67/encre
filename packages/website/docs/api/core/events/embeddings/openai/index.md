# events/embeddings/openai

## Index

### Classes

- [OpenAIEmbeddings](classes/OpenAIEmbeddings.md)

### Interfaces

- [OpenAIEmbeddingsCallOptions](interfaces/OpenAIEmbeddingsCallOptions.md)
- [OpenAIEmbeddingsParams](interfaces/OpenAIEmbeddingsParams.md)
