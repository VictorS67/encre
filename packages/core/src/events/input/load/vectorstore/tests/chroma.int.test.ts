import { describe, expect, test } from '@jest/globals';
import { IncludeEnum } from 'chromadb';
import { OpenAIEmbeddings } from '../../../../embeddings/openai.js';
import { Context } from '../../docs/context.js';
import { ChromaVectorStore } from '../chroma.js';

describe('ChromaVectorStore', () => {
  /**
   * Before testing, deploy Chroma on a long-running server, and connect to it remotely
   *
   * Docker:
   * ```
   * docker pull chromadb/chroma
   * docker run -p 8000:8000 chromadb/chroma
   * ```
   */

  const OPENAI_API_KEY = 'you_should_get_this_api_from_openai';

  const openaiEmbeddings = new OpenAIEmbeddings({
    openAIApiKey: OPENAI_API_KEY,
    modelName: 'text-embedding-ada-002',
  });

  const embed = async (query: string): Promise<number[]> => {
    const embedResult = await openaiEmbeddings.invoke(query);

    return embedResult.embedding;
  };

  test('imports correctly', async () => {
    const { ChromaClient } = await import('chromadb');

    expect(ChromaClient).toBeDefined();
  });

  test('constructor defined', async () => {
    const chromaStore = new ChromaVectorStore({
      url: 'http://localhost:8000',
      collectionName: 'test-collection',
    });

    expect(chromaStore).toBeDefined();
  });

  test('similarity search', async () => {
    const chroma = new ChromaVectorStore({
      url: 'http://localhost:8000',
      collectionName: 'test-similarity-search',
    });

    chroma.addVectors(
      [
        await embed('I like pizza'),
        await embed("I don't like dogs"),
        await embed('I looks like a clown'),
      ],
      [
        new Context({
          pageContent: 'I like pizza',
        }),
        new Context({
          pageContent: "I don't like dogs",
        }),
        new Context({
          pageContent: 'I looks like a clown',
        }),
      ],
      {
        ids: ['vec1', 'vec2', 'vec3'],
      }
    );

    const searches = await chroma.similaritySearch(
      await embed('what do I like?'),
      2
    );

    expect(searches).toMatchSnapshot();
  });

  test('mmr search', async () => {
    const chroma = new ChromaVectorStore({
      url: 'http://localhost:8000',
      collectionName: 'test-mmr-search',
    });

    chroma.addVectors(
      [
        await embed('I like pizza'),
        await embed("I don't like dogs"),
        await embed('I looks like a clown'),
      ],
      [
        new Context({
          pageContent: 'I like pizza',
        }),
        new Context({
          pageContent: "I don't like dogs",
        }),
        new Context({
          pageContent: 'I looks like a clown',
        }),
      ],
      {
        ids: ['vec1', 'vec2', 'vec3'],
      }
    );

    const searches = await chroma.maxMarginalRelevanceSearch(
      await embed('what do I like?'),
      2,
      0.5
    );

    expect(searches).toMatchSnapshot();
  });

  test('add vectors to the collection with loc metadata', async () => {
    const expectedVectorIds: string[] = ['vec1', 'vec2', 'vec3'];

    const expectedPageContents: string[] = [
      'I like pizza',
      "I don't like dogs",
      'I looks like a clown',
    ];

    const expectedEmbeddings: number[][] = await Promise.all(
      expectedPageContents.map(async (pc) => await embed(pc))
    );

    const expectedContexts: Context[] = expectedPageContents.map(
      (pc, index) =>
        new Context({
          pageContent: pc,
          metadata: {
            source: 'dummy.html',
            loc: {
              lines: {
                from: 0 + index * 4,
                to: (index + 1) * 4,
              },
            },
          },
        })
    );

    const expectedMetadatas = expectedPageContents.map((_, index) => ({
      source: 'dummy.html',
      locFrom: 0 + index * 4,
      locTo: (index + 1) * 4,
    }));

    const chromaStore = new ChromaVectorStore({
      url: 'http://localhost:8000',
      collectionName: 'test-add-vectors',
    });

    chromaStore.addVectors(expectedEmbeddings, expectedContexts, {
      ids: expectedVectorIds,
    });

    const collection = await chromaStore._getCollection();

    const response = await collection.get({
      ids: expectedVectorIds,
      include: [IncludeEnum.Metadatas],
    });

    expect(JSON.parse(JSON.stringify(response.metadatas))).toStrictEqual(
      expectedMetadatas
    );
  });
});
