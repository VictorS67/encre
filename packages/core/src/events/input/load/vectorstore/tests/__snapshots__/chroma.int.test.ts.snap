// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromaVectorStore mmr search 1`] = `
[
  [
    Context {
      "metadata": {},
      "pageContent": "I like pizza",
    },
    0.833390357575667,
  ],
  [
    Context {
      "metadata": {},
      "pageContent": "I don't like dogs",
    },
    -0.014507155088069712,
  ],
]
`;

exports[`ChromaVectorStore similarity search 1`] = `
[
  [
    Context {
      "metadata": {},
      "pageContent": "I like pizza",
    },
    0.33321927209466545,
  ],
  [
    Context {
      "metadata": {},
      "pageContent": "I don't like dogs",
    },
    0.44407291953459693,
  ],
]
`;
